ast/compile.go

```diff
@@ -4165,66 +4232,84 @@ func rewriteDeclaredVarsInTerm(g *localVarGenerator, stack *localDeclaredVars, t
 	case *object:
 		cpy, _ := v.Map(func(k, v *Term) (*Term, *Term, error) {
 			kcpy := k.Copy()
-			errs = rewriteDeclaredVarsInTermRecursive(g, stack, kcpy, errs)
-			errs = rewriteDeclaredVarsInTermRecursive(g, stack, v, errs)
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, kcpy, errs, strict)
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, v, errs, strict)
 			return kcpy, v, nil
 		})
 		term.Value = cpy
 	case Set:
 		cpy, _ := v.Map(func(elem *Term) (*Term, error) {
 			elemcpy := elem.Copy()
-			errs = rewriteDeclaredVarsInTermRecursive(g, stack, elemcpy, errs)
+			errs = rewriteDeclaredVarsInTermRecursive(g, stack, elemcpy, errs, strict)
 			return elemcpy, nil
 		})
 		term.Value = cpy
 	case *ArrayComprehension:
-		errs = rewriteDeclaredVarsInArrayComprehension(g, stack, v, errs)
+		errs = rewriteDeclaredVarsInArrayComprehension(g, stack, v, errs, strict)
 	case *SetComprehension:
-		errs = rewriteDeclaredVarsInSetComprehension(g, stack, v, errs)
+		errs = rewriteDeclaredVarsInSetComprehension(g, stack, v, errs, strict)
 	case *ObjectComprehension:
-		errs = rewriteDeclaredVarsInObjectComprehension(g, stack, v, errs)
+		errs = rewriteDeclaredVarsInObjectComprehension(g, stack, v, errs, strict)
 	default:
 		return false, errs
 	}
 	return true, errs
 }
 
-func rewriteDeclaredVarsInTermRecursive(g *localVarGenerator, stack *localDeclaredVars, term *Term, errs Errors) Errors {
+func rewriteDeclaredVarsInTermRecursive(g *localVarGenerator, stack *localDeclaredVars, term *Term, errs Errors, strict bool) Errors {
 	WalkNodes(term, func(n Node) bool {
 		var stop bool
 		switch n := n.(type) {
 		case *With:
-			_, errs = rewriteDeclaredVarsInTerm(g, stack, n.Value, errs)
+			_, errs = rewriteDeclaredVarsInTerm(g, stack, n.Value, errs, strict)
 			stop = true
 		case *Term:
-			stop, errs = rewriteDeclaredVarsInTerm(g, stack, n, errs)
+			stop, errs = rewriteDeclaredVarsInTerm(g, stack, n, errs, strict)
 		}
 		return stop
 	})
 	return errs
 }
 
-func rewriteDeclaredVarsInArrayComprehension(g *localVarGenerator, stack *localDeclaredVars, v *ArrayComprehension, errs Errors) Errors {
+func rewriteDeclaredVarsInArrayComprehension(g *localVarGenerator, stack *localDeclaredVars, v *ArrayComprehension, errs Errors, strict bool) Errors {
+	used := NewVarSet()
+	if v.Term != nil {
+		used.Update(v.Term.Vars())
+	}
```

Comment:
This will change the error generated by `p := [i | some i]` from `declared var i unused` to `var i is unsafe`.
set and object comprehension is similarly changed.

Reply:
Can these keys, and the object pairs, ever be nil? I might just be missing something 🤔

Reply:
I don't think so, just playing it safe.

Reply:
I don't think so, just playing it safe.

Reply:
I think this is muddying the waters a bit. But it's not a blocker, of course.

Reply:
Did some extra checking, and indeed it looks safe to not do any nil-checking.